Revert CI back to `macos-latest`

Xophmeister opened the pull request:
# Revert CI back to `macos-latest`

<!----------------------------------------------------------------------
If this PR is related to, or resolves an issue, please reference it
here.

- For issues that are fixed by this PR, use GitHub's automatic closing
  feature by writing "Resolves #XXX", for example.

- If multiple issues are implicated, please list them out, with an
  appropriate byline for each issue, with one issue per line.

- You may also use this section to reference other PRs and discussion
  items, following the same pattern outlined here for issues.

Below is an example; please update as appropriate. If no issue, etc. is
implicated in this PR, please remove this section.
----------------------------------------------------------------------->
Resolves #1043
References https://github.com/actions/runner-images/issues/12588

## Description

Reverts the main CI's macOS run to use GitHub's `macos-latest` runner, rather than `macos-14`, as this has been fixed upstream.

## Checklist

<!----------------------------------------------------------------------
See MAINTAINERS.md for more details.
This is particularly important if this PR is preparing a release.
----------------------------------------------------------------------->

Checklist before merging, wherever relevant:

- [ ] ~`CHANGELOG.md` updated~
- [x] Documentation (The Topiary Book, `README.md`, etc.) up-to-date
<!----------------------------------------------------------------------
If the PR solves a formatting issue for a supported language,
or generally improves formatting, please make sure to include a
regression test showcasing the fix/improvement to:
`topiary-cli/tests/samples/{input,expected}/mylanguage.lang`
----------------------------------------------------------------------->
- [ ] ~Updated regression tests~

avdv commented:
Hi, I am wondering... macos-14 and macos-latest are the same thing currently. (migration to macos-15 will start in August)

Xophmeister commented:
> Hi, I am wondering... macos-14 and macos-latest are the same thing currently. (migration to macos-15 will start in August)

I didn't actually check, but that's the same thing GitHub said. However, the build was failing and there was an error about the new macOS runners, so I changed it to `macos-14` and that fixed the problem. Then GitHub told me the same thing as you, so I reverted it and the build still works... So I'm super-confused :shrug:

Maybe the build broke for a completely unrelated reason, due to some other action, which was fixed very quickly in between the time of me noticing it and making a fix (less than a day). Or perhaps Topiary was subjected to random A/B testing and we got unlucky. I don't know what's the simplest explanation is here :sweat_smile:

Either way, it's working now, so maybe my curiosity doesn't need to be satisfied!
